pkg/rhos/securitygroups.go

```diff
@@ -0,0 +1,244 @@
+	opts := servers.ListOpts{Name: nodeName}
+	pager := servers.List(computeClient, opts)
+
+	err := pager.EachPage(func(page pagination.Page) (bool, error) {
+		serverList, err := servers.ExtractServers(page)
+		if err != nil {
+			return false, errors.WithMessagef(err, "getting the server List failed for node %q", nodeName)
+		}
+		for i := range serverList {
+			err = secgroups.AddServer(computeClient, serverList[i].ID, groupName).ExtractErr()
+			if err != nil {
+				return false, errors.WithMessagef(err, "adding security group %q to the server %q failed",
+					groupName, serverList[i].Name)
+			}
+		}
+
+		return true, nil
+	})
+
+	return errors.WithMessagef(err, "open gateway ports failed")
+}
+
+func (c *CloudInfo) removeGWFirewallRules(groupName, nodeName string, computeClient *gophercloud.ServiceClient) error {
+	opts := servers.ListOpts{Name: nodeName}
+	pager := servers.List(computeClient, opts)
+
+	err := pager.EachPage(func(page pagination.Page) (bool, error) {
+		serverList, err := servers.ExtractServers(page)
+		if err != nil {
+			return false, errors.WithMessagef(err, "getting the server list failed")
+		}
+		for i := range serverList {
+			err = secgroups.RemoveServer(computeClient, serverList[i].ID, groupName).ExtractErr()
+			if err != nil {
+				return false, errors.WithMessagef(err, "failed to remove the firewall for the server: %q", serverList[i].Name)
+			}
+		}
+
+		return true, nil
+	})
+
+	return errors.WithMessagef(err, "removing firewall rules failed for security group %q", groupName)
+}
+
+func (c *CloudInfo) deleteSG(groupName string, computeClient *gophercloud.ServiceClient) error {
+	pager := secgroups.List(computeClient)
+	var isFound bool
+	var securityGroupID string
+
+	err := pager.EachPage(func(page pagination.Page) (bool, error) {
+		serverList, err := secgroups.ExtractSecurityGroups(page)
+		if err != nil {
+			return false, errors.WithMessagef(err, "failed to list the security group %q", groupName)
+		}
+		for _, s := range serverList {
+			if s.Name == groupName {
+				isFound = true
+				securityGroupID = s.ID
+			}
+		}
+		return true, errors.WithMessagef(err, "error finding the uuid for the security group: %q", groupName)
+	})
+
+	if err != nil && isFound {
+		err = secgroups.Delete(computeClient, securityGroupID).ExtractErr()
+	}
+
+	return errors.WithMessagef(err, "error deleting the security group %q", groupName)
+}
+
+func (c *CloudInfo) createSGRule(group, remoteGroupID, remoteIPPrefix string, port uint16,
+	protocol string, networkClient *gophercloud.ServiceClient) error {
+	opts := rules.CreateOpts{
+		Direction:      "ingress",
+		EtherType:      rules.EtherType4,
+		SecGroupID:     group,
+		PortRangeMax:   int(port),
+		PortRangeMin:   int(port),
+		Protocol:       rules.RuleProtocol(protocol),
+		RemoteGroupID:  remoteGroupID,
+		RemoteIPPrefix: remoteIPPrefix,
+	}
+
+	_, err := rules.Create(networkClient, opts).Extract()
```

Comment:
Should we check if the error is because the SG rule already exists?

Reply:
we always check if the security group exists and skips creation if it already exists So it is safe to assume if the security rules will not be present right?